domain/controllerconfig/service/service.go

```diff
@@ -174,15 +174,30 @@ func validObjectStoreProgression(current map[string]string, updateAttrs controll
 		return errors.Errorf("can not remove %q", controller.ObjectStoreType)
 	}
 
+	// If we're not changing the object store type, we don't need to validate
+	// anything.
+	if _, ok := updateAttrs[controller.ObjectStoreType]; !ok {
+		return nil
+	}
+
 	// We should always have a valid object store type in the current config,
 	// so we don't need to check for errors.
 	cur := objectstore.BackendType(current[controller.ObjectStoreType])
 	upd := updateAttrs.ObjectStoreType()
 
 	// We're not changing the object store type, or we're changing from
 	// filestorage to s3storage.
-	if cur == upd || cur == objectstore.FileBackend && upd == objectstore.S3Backend {
+	if cur == upd {
 		return nil
+	} else if cur == objectstore.FileBackend && upd == objectstore.S3Backend {
+		// To be 100% sure that we can change from filestorage to s3storage,
+		// we're going to check if the current config has a complete s3 config.
+		// This is rather expensive, but it's the only way to be sure that we
+		// can change from filestorage to s3storage.
+		if controller.HasCompleteS3ControllerConfig(updateAttrs) || hashCompleteS3Config(current) {
```

Comment:
You could make this return nil or an error and give more detail as to what's wrong.